src/http-request.ts

```diff
@@ -31,3 +45,194 @@ export interface CloudLoggingHttpRequest {
+    const url = new URL(requestUrl);
+    protocol = url.protocol;
+  }
+  req.method ? (requestMethod = req.method) : null;
+  if (req.headers && req.headers['user-agent']) {
+    req.headers['user-agent'] ? (userAgent = req.headers['user-agent']) : null;
+    req.headers['referer'] ? (referer = req.headers['referer']) : null;
+  }
+  // Format response properties
+  if (res) {
+    res.statusCode ? (status = res.statusCode) : null;
+    responseSize =
+      (res.getHeader && Number(res.getHeader('Content-Length'))) || 0;
+  }
+  // Format latency
+  if (latencyMilliseconds) {
+    latency = {
+      seconds: Math.floor(latencyMilliseconds / 1e3),
+      nanos: Math.floor((latencyMilliseconds % 1e3) * 1e6),
+    };
+  }
+  // Only include the property if its value exists
+  return Object.assign(
+    {},
+    requestUrl ? {requestUrl} : null,
+    protocol ? {protocol} : null,
+    requestMethod ? {requestMethod} : null,
+    userAgent ? {userAgent} : null,
+    referer ? {referer} : null,
+    responseSize ? {responseSize} : null,
+    status ? {status} : null,
+    latency ? {latency} : null
+  );
+}
+
+/**
+ * HeaderWrapper: wraps getHeader and setHeader operations.
+ */
+export type HeaderWrapper = w3cContext.HeaderGetter & w3cContext.HeaderSetter;
+
+/**
+ * makeHeaderWrapper returns a wrapper with set and get header functionality,
+ * returning null if the incoming request object doesn't contain the 'header'
+ * propery.
+ * @param req
+ */
+export function makeHeaderWrapper(
+  req: http.IncomingMessage
+): HeaderWrapper | null {
+  if (!req.headers) return null;
+  const wrapper = {
+    setHeader(name: string, value: string) {
+      req.headers[name] = value;
+    },
+    getHeader(name: string) {
+      return req.headers[name];
+    },
+  };
+  return wrapper;
+}
+
+export interface CloudTraceContext {
+  trace: string;
+  spanId?: string;
+  traceSampled?: boolean;
+}
+
+/**
+ * getTraceContext returns a CloudTraceContext with as many available trace and
+ * span properties as possible. It examines HTTP headers for trace context.
+ * Optionally, it can inject a W3C compliant trace context when no context is
+ * available from headers.
+ */
+export function getTraceContext(
+  req: http.IncomingMessage,
+  projectId: string,
+  inject?: boolean
+): CloudTraceContext {
+  const context: CloudTraceContext = {
+    trace: '',
+    spanId: undefined,
+    traceSampled: undefined,
+  };
+  const wrapper = makeHeaderWrapper(req);
+  if (wrapper) {
+    // Detect 'X-Cloud-Trace-Context' header.
+    const cloudContext = getCloudTraceContext(wrapper);
+    if (cloudContext) {
+      context.trace = `projects/${projectId}/traces/${cloudContext.trace}`;
+      context.spanId = cloudContext.spanId;
+      context.traceSampled = cloudContext.traceSampled;
+    } else {
+      // Detect 'traceparent' header, optionally injecting one if not available.
+      const parentContext = getOrInjectTraceParent(wrapper, inject);
```

Comment:
Perhaps, worth extracting the entire trace/span detection and propagation part into a separate method. It will also make it easier to test.

Reply:
that's a good suggestion! will do